src/ui/view-page.tsx

```diff
@@ -0,0 +1,535 @@
+            class: "cm-punctuation",
+        },
+    ])
+);
+const LANG_COMPARTMENT = new Compartment();
+const EDITOR_EXTS = [
+    lineNumbers(),
+    highlightSpecialChars(),
+    history(),
+    foldGutter(),
+    drawSelection(),
+    dropCursor(),
+    EditorState.allowMultipleSelections.of(true),
+    indentOnInput(),
+    bracketMatching(),
+    closeBrackets(),
+    rectangularSelection(),
+    crosshairCursor(),
+    highlightSelectionMatches(),
+    keymap.of([
+        indentWithTab,
+        ...closeBracketsKeymap,
+        ...defaultKeymap,
+        ...searchKeymap,
+        ...historyKeymap,
+        ...foldKeymap,
+        ...completionKeymap,
+        ...lintKeymap,
+    ]),
+    EditorView.baseTheme({
+        ".cm-cursor": {
+            borderLeftColor: "var(--text)",
+        },
+        ".cm-tooltip": {
+            backgroundColor: "var(--bg)",
+        },
+        "&:not(.cm-focused) .cm-fat-cursor": {
+            outline: "solid 1px var(--accent) !important",
+            background: "none",
+        },
+        "&.cm-focused .cm-fat-cursor": {
+            background: "var(--accent) !important",
+        },
+    }),
+];
+/** Provides a minimal editor with syntax highlighting */
+function CodeMirrorEditor({
+    state: { script, sourceType: lang },
+    setState,
+}: {
+    lang?: ScriptLanguage;
+    state: DatacoreViewState;
+    setState: (state: Partial<DatacoreViewState>) => void;
+}) {
+    const editorRef = useRef<HTMLDivElement>(null);
+    const viewRef = useRef<EditorView>(null);
+    const viewContext = useContext(CUSTOM_VIEW_CONTEXT);
+    useEffect(() => {
+        if (editorRef.current && !viewRef.current) {
+            viewRef.current = new EditorView({
+                parent: editorRef.current,
+                extensions: [viewContext.app.vault.getConfig("vimMode") && vim()].concat(
+                    ...EDITOR_EXTS.concat(
+                        ...[
+                            LANG_COMPARTMENT.of(javascript()),
+                            ViewPlugin.fromClass(
+                                class {
+                                    constructor(public view: EditorView) {}
+                                    update(update: ViewUpdate) {
+                                        if (update.docChanged) {
+                                            setState({ script: this.view.state.sliceDoc() || "" });
+                                        }
+                                    }
+                                }
+                            ),
+                            EDITOR_HL,
+                            viewContext.app.vault.getConfig("vimMode") && vim(),
+                            ...(viewContext.app.plugins.plugins["datacore-addon-autocomplete"]?.extensions || []),
+                        ].filter((a) => !!a)
+                    )
+                ),
+                doc: script || "",
+            });
+        }
+        return () => viewRef.current?.destroy();
```

Comment:
I'm a big fan of short circuiting patterns when possible. :) I've not verified it, but something like below makes a lot of sense to my brain and helps prevent getting if statments too nested as the code grows.

```suggestion
        if (!editorRef.current || viewRef.current) {
            return () => viewRef.current?.destroy();
        }
        viewRef.current = new EditorView({
            parent: editorRef.current,
            extensions: [viewContext.app.vault.getConfig("vimMode") && vim()].concat(
                ...EDITOR_EXTS.concat(
                    ...[
                        LANG_COMPARTMENT.of(javascript()),
                        ViewPlugin.fromClass(
                            class {
                                constructor(public view: EditorView) {}
                                update(update: ViewUpdate) {
                                    if (update.docChanged) {
                                        setState({ script: this.view.state.sliceDoc() || "" });
                                    }
                                }
                            }
                        ),
                        EDITOR_HL,
                        viewContext.app.vault.getConfig("vimMode") && vim(),
                        ...(viewContext.app.plugins.plugins["datacore-addon-autocomplete"]?.extensions || []),
                    ].filter((a) => !!a)
                )
            ),
            doc: script || "",
        });
        return () => viewRef.current?.destroy();
```

Reply:
why are you returning the useEffect cleanup early?

Reply:
I could be missing a nuance for how useEffect clean up functions work, but my main goal here was to `early exit` or `short circuit` the if statement. Since the current function always returned the clean up code, I wanted my modified suggestion also to return it. This article does a pretty good job describing it: https://medium.com/swlh/return-early-pattern-3d18a41bba8

Arguably since there is only one `if` statement, it might not be truly necessary yet, but I like to start the pattern early so that you don't end up with a bunch of nested `if` statements as the code base invariably grows. But some people might argue that's `over engineering`. Just thought I'd bring it up for discussion. 🙂